pubsub_nats_test.go

```diff
@@ -578,6 +578,62 @@ func TestSubscribe(t *testing.T) {
 
 			},
 		},
+		{
+			description: "should receive error in errors channel if you take too long to respond to publication",
+			setup: func(t *testing.T, pub *Publication, _ PubSubClient) {
+				// high level overview of this test:
+				//   - make a request expecting to get back a publication
+				//   - subscriber gets the publication
+				//   - subscriber responds to the publication via the Reply call
+				//   - the Reply call will fail because the replyCh will be set to nil by the message
+				//     handler as the deadline will be reached.
+				//   - this should result in an error being sent to the configured errors channel
+
+				// act as a client - publish a message expecting to get back a Publication response
+				pub.ResponseMode = ResponseModePublication
+				data, err := elemental.Encode(elemental.EncodingTypeMSGPACK, pub)
+				if err != nil {
+					t.Fatalf("test setup failed - could not encode publication - error: %+v", err)
+					return
+				}
+
+				// our request should fail as the subscriber is not going to be able to respond in time due to an artificial
+				// deadline (0 nanoseconds) - see the subscribe options for this test
+				response, err := nc.Request(subscribeTopic, data, threshold)
+				if err == nil {
+					t.Fatalf("test setup failed - expected to get an error here, but got a response back instead: \"%+v\". "+
+						"Hint: test structure has likely been messed up.", *response)
+					return
+				}
+			},
+			// we expect to get an error back, because we take longer than 0 nanoseconds to respond to the received publication :-)
+			expectedError: errors.New(""),
+			subscribeOptions: []PubSubOptSubscribe{
+				// we deliberately set a low timeout here!
+				NATSOptSubscribeReplyTimeout(0),
```

Comment:
0 should mean "no timeout"

Reply:
that also mean that as a user of the api I should be able to Abort waiting for reply, which should stop the PubSubClient to wait for a reponse

Reply:
> that also means that as a user of the api I should be able to Abort waiting for reply, which should stop the PubSubClient to wait for a response

You can.

If you are publishing a message, you have the option to provide a context in your call to `Publish` via:

```go
func NATSOptRespondToChannel(ctx context.Context, resp chan *Publication) PubSubOptPublish
```
```go
func NATSOptPublishRequireAck(ctx context.Context) PubSubOptPublish
```

Reply:
ok nice. then just "0 should mean no timeout" :)

Reply:
I was following the advice mentioned here https://dave.cheney.net/practical-go/presentations/qcon-china.html#_never_start_a_goroutine_without_knowning_when_it_will_stop

I don't think it's wise to let someone cause an intentional resource leak by never timing out. The goroutine that is handling the response channel will never return if someone does not call `Reply` otherwise.